components/node-labeler/cmd/run.go

```diff
@@ -197,90 +224,183 @@ func (r *PodReconciler) Reconcile(ctx context.Context, req reconcile.Request) (r
+
+	return reconcile.Result{}, nil
+}
+
+func checkPodHealth(pod corev1.Pod) (bool, error) {
 	var (
-		ipAddress     string
-		port          string
-		component     string
-		labelToUpdate string
+		ipAddress string
+		port      string
 	)
-
 	switch {
 	case strings.HasPrefix(pod.Name, registryFacade):
-		component = registryFacade
-		labelToUpdate = fmt.Sprintf(registryFacadeLabel, namespace)
 		ipAddress = pod.Status.HostIP
 		port = strconv.Itoa(registryFacadePort)
 	case strings.HasPrefix(pod.Name, wsDaemon):
-		component = wsDaemon
-		labelToUpdate = fmt.Sprintf(wsdaemonLabel, namespace)
 		ipAddress = pod.Status.PodIP
 		port = strconv.Itoa(wsdaemonPort)
 	default:
 		// nothing to do
-		return reconcile.Result{}, nil
+		return true, nil
 	}
 
 	if !pod.ObjectMeta.DeletionTimestamp.IsZero() {
 		// the pod is being removed.
-		// remove the component label from the node
-		time.Sleep(1 * time.Second)
-		err := updateLabel(labelToUpdate, false, nodeName, r)
-		if err != nil {
-			// this is a edge case when cluster-autoscaler removes a node
-			// (all the running pods will be removed after that)
-			if errors.IsNotFound(err) {
-				return reconcile.Result{}, nil
-			}
-
-			log.WithError(err).Error("removing node label")
-			return reconcile.Result{RequeueAfter: defaultRequeueTime}, err
-		}
-
-		return reconcile.Result{}, err
+		// add the taint to the node
+		return false, nil
 	}
 
 	if !IsPodReady(pod) {
 		// not ready. Wait until the next update.
-		return reconcile.Result{}, nil
-	}
-
-	var node corev1.Node
-	err = r.Get(ctx, types.NamespacedName{Name: nodeName}, &node)
-	if err != nil {
-		return reconcile.Result{}, fmt.Errorf("obtaining node %s: %w", nodeName, err)
-	}
-
-	if labelValue, exists := node.Labels[labelToUpdate]; exists && labelValue == "true" {
-		// nothing to do, the label already exists.
-		return reconcile.Result{}, nil
+		return false, nil
 	}
 
-	err = checkTCPPortIsReachable(ipAddress, port)
+	err := checkTCPPortIsReachable(ipAddress, port)
 	if err != nil {
 		log.WithField("host", ipAddress).WithField("port", port).WithField("pod", pod.Name).WithError(err).Error("checking if TCP port is open")
-		return reconcile.Result{RequeueAfter: defaultRequeueTime}, nil
+		return false, nil
 	}
 
-	if component == registryFacade {
+	if strings.HasPrefix(pod.Name, registryFacade) {
 		err = checkRegistryFacade(ipAddress, port)
 		if err != nil {
 			log.WithError(err).Error("checking registry-facade")
-			return reconcile.Result{RequeueAfter: defaultRequeueTime}, nil
+			return false, nil
 		}
+	}
+
+	return true, nil
+}
 
-		time.Sleep(1 * time.Second)
+type NodeReconciler struct {
+	client.Client
+}
+
+func (r *NodeReconciler) nodeFilter() predicate.Predicate {
+	return predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			node, ok := e.Object.(*corev1.Node)
+			if !ok {
+				return false
+			}
+			return isWorkspaceNode(*node)
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			return false
+		},
```

Comment:
How come it isn't necessary to do on update, too? 

I ask because it was necessary for CA autoscaler, to prevent data loss for workspaces having a large backup taking a long time, to do on create and update of workspace CRD. (ref: https://github.com/gitpod-io/gitpod/blob/1d0781019ec275f07b358a237e5ab802d347c3f1/components/node-labeler/cmd/run.go#L486)

Reply:
It seems like it is unnecessary, because we aren't working with a potential nil (like runtime for workspace).

Reply:
Ah, right, we do reconcileAll here: https://github.com/gitpod-io/gitpod/blob/1d0781019ec275f07b358a237e5ab802d347c3f1/components/node-labeler/cmd/run.go#L131